pytential/symbolic/pde/system_utils.py

```diff
@@ -0,0 +1,744 @@
+       * If there is a sum of two `IntG`s with same target derivative and different
+         source derivatives of the same kernel, they are merged into one FMM.
+
+       * If possible, convert :class:`sumpy.kernel.AxisSourceDerivative` to
+         :class:`sumpy.kernel.DirectionalSourceDerivative`.
+
+       * Reduce the number of FMMs by converting the `IntG` expression to
+         a matrix and factoring the matrix where the left operand matrix represents
+         a transformation at target and the right matrix represents a transformation
+         at source. For this to work, we need to know which variables depend on
+         source so that they do not end up in the left operand. User needs to supply
+         this as the argument `source_dependent_variable`.
+
+    :arg base_kernel: A :class:`sumpy.kernel.Kernel` object if given will be used
+        for converting a :class:`~pytential.symbolic.primitives.IntG` to a linear
+        expression of same type with the kernel replaced by base_kernel and its
+        derivatives
+
+    :arg verbose: increase verbosity of merging
+
+    :arg source_dependent_variable: When merging expressions, consider only these
+        variables as dependent on source. Otherwise consider all variables
+        as source dependent. This is important when reducing the number of FMMs
+        needed for the output.
+    """
+
+    int_g_s = get_int_g_s(exprs)
+
+    if base_kernel is not None:
+        replacements = {}
+
+        # Iterate all the IntGs in the expressions and create a dictionary
+        # of replacements for the IntGs
+        for int_g in int_g_s:
+            # First convert IntGs with target derivatives to source derivatives
+            new_int_g = convert_target_deriv_to_source(int_g)
+            # Convert IntGs with TargetMultiplier to a sum of IntGs without
+            # TargetMultipliers
+            new_int_g_s = convert_target_multiplier_to_source(new_int_g)
+            # Convert IntGs with different kernels to expressions containing
+            # IntGs with base_kernel or its derivatives
+            replacements[int_g] = sum(convert_int_g_to_base(new_int_g,
+                base_kernel, verbose=verbose) for new_int_g in new_int_g_s)
+
+        # replace the IntGs in the expressions
+        substitutor = IntGSubstitutor(replacements)
+        exprs = [substitutor(expr) for expr in exprs]
+
+    differing_kernel_arguments = set()
+    kernel_arguments = {}
+    for int_g in int_g_s:
+        for k, v in int_g.kernel_arguments.items():
+            if k not in kernel_arguments:
+                kernel_arguments[k] = v
+            else:
+                if kernel_arguments[k] != v:
+                    differing_kernel_arguments.add(k)
+
+    # Using a dictionary instead of a set because sets are unordered
+    all_source_groups = dict()
+
+    result = np.array([0 for _ in exprs], dtype=object)
+
+    int_gs_by_group_for_index = []
+
+    int_g_cc = IntGCoefficientCollector()
```

Comment:
:warning: Think about how this works with nested layer potentials (in future review)